feat(code-first): Use capitalized case for enums in code-first

kasvith opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/nestjs/nest/blob/master/CONTRIBUTING.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: #1832 


## What is the new behavior?
`registerEnumType` now contains an optional parameter called `mapToUppercase`(defaults to `false` to avoid breaking changes). This option automatically converts enum to `CONST_CASE` in compliance with GraphQL best practices.
It means we can use `PascalCase` in `typescript` while using `CONST_CASE` in GraphQL.

```ts
export enum CatType {
  PersianCat = 'persian-cat',
  MaineCoon = 'maine-coon',
  Ragdoll = 'ragdoll',
}

registerEnumType(CatType, {
  name: 'CatType',
  description: 'Distinguish cats',
  mapToUppercase: true, 
});
```

Would produce the following enum type in schema
```gql
"""Distinguish cats"""
enum CatType {
  PERSIAN_CAT
  MAINE_COON
  RAGDOLL
}
```

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

Since this is a norm in GraphQL world, from the next release I consider we should make `mapToUppercase` to `true` by default. So people can use natural `PascalCase` enum values in `typescript` and use `CONST_CASE` by default in `GraphQL`

kasvith committed:
feat(): added enum case mapper

kasvith committed:
feat(): added enum metadata

kasvith committed:
feat(): added register enum type factory option

kasvith committed:
refactor: change name of variable

kasvith commented:
@kamilmysliwiec please take a look :)

kasvith commented:
@kamilmysliwiec would you please take a look at this

kasvith commented:
I added docs in `JSDoc` format for the added option. But I would also add this to nest `docs` once its merged

micalevisk commented on lib/schema-builder/factories/enum-definition.factory.ts:
```diff
@@ -1,7 +1,11 @@
 import { EnumMetadata } from '../metadata';
 
+export const mapToUppercase = (value: string): string =>
+  upperCase(value).replace(' ', '_');
```
is this expected to replace only the first blank space?

kasvith replied:
oh, true it should convert all blank spaces. I will fix it. Good catch

kasvith replied:
fixed it

kasvith committed:
fix(code-first): use replaceAll for uppercase mapper

kasvith commented:
Hi, @kamilmysliwiec I've fixed the merge conflicts. Some tests are failing related to the gateway module. Tests I added are passing in e2e but the build fails due to mentioned issues. How can we proceed?

kasvith committed:
fix(code-first): respect new apollo loader

kasvith committed:
refactor(code-first): rearrange import

kasvith commented:
I've resolved the merge conflict. Also addressed comments by @micalevisk.
Can you take a look @kamilmysliwiec ? 😃

kasvith committed:
test(): added tests for enum capitalization

kasvith committed:
test(): added tests for enum capitalization

kasvith committed:
test(code-first): added tests for enum capitalization

kasvith committed:
test(code-first): added more testcases to enum case mapper

kasvith commented:
I added more test cases to the feature

kasvith commented:
Hello @kamilmysliwiec is there any improvement that I need to do in this PR?

kasvith commented:
Hi @kamilmysliwiec would you be able to review this PR and give me a comment about the future of it :) thanks

peshkov3 commented:
Waiting this PR

kasvith commented:
Im ready to resolve conflicts once it gets proper attention

kasvith commented:
@kamilmysliwiec would you be able to review this and get the approval

i can resolve the merge conflict afterwards

kasvith commented:
Hi @kamilmysliwiec any news about this?

mareksuscak commented on packages/graphql/lib/schema-builder/metadata/enum.metadata.ts:
```diff
@@ -12,4 +12,5 @@ export interface EnumMetadata<T extends object = any> {
   name: string;
   description?: string;
   valuesMap?: EnumMetadataValuesMap<T>;
+  mapToUppercase?: boolean;
```
@kasvith this option seems to be a single purpose one. Would it be better to take a key transform function so that people can use any function they want for the enum keys? toUppercase could then become the default and the feature would become a lot more useful.

kasvith replied:
Hi @mareksuscak

By convention GraphQL enums are UPPER_CASE literals, i dont see a usecase to have a transformer function and make this much more complicated than it is now

https://www.apollographql.com/tutorials/side-quest-intermediate-schema-design/02-the-enum-type
